snakemake/executors.py

```diff
@@ -2052,3 +2059,605 @@ def run_wrapper(
+            restart_times=restart_times,
+            exec_job=exec_job,
+            assume_shared_fs=False,
+            max_status_checks_per_second=10,
+        )
+
+    def _get_services(self):
+        """use the Google Discovery Build to generate API clients
+           for Life Sciences, and use the google storage python client 
+           for storage.
+        """
+        from googleapiclient.discovery import build as discovery_build
+        from oauth2client.client import (
+            GoogleCredentials,
+            ApplicationDefaultCredentialsError,
+        )
+        from google.cloud import storage
+
+        # Credentials must be exported to environment
+        try:
+            creds = GoogleCredentials.get_application_default()
+        except ApplicationDefaultCredentialsError as ex:
+            log_verbose_traceback(ex)
+            raise ex
+
+        # Discovery clients for Google Cloud Storage and Life Sciences API
+        self._storage_cli = discovery_build("storage", "v1", credentials=creds)
+        self._compute_cli = discovery_build("compute", "v1", credentials=creds)
+        self._api = discovery_build("lifesciences", "v2beta", credentials=creds)
+        self._bucket_service = storage.Client()
+
+    def _get_bucket(self):
+        """get a connection to the storage bucket (self.bucket) and exit
+           if the name is taken or otherwise invalid.
+
+           Parameters
+           ==========
+           workflow: the workflow object to derive the prefix from
+        """
+        # Hold path to requested subdirectory and main bucket
+        bucket_name = self.workflow.default_remote_prefix.split("/")[0]
+        self.gs_subdir = re.sub(
+            "^{}/".format(bucket_name), "", self.workflow.default_remote_prefix
+        )
+
+        # Case 1: The bucket already exists
+        try:
+            self.bucket = self._bucket_service.get_bucket(bucket_name)
+
+        # Case 2: The bucket needs to be created
+        except google.cloud.exceptions.NotFound:
+            self.bucket = self._bucket_service.create_bucket(bucket_name)
+
+        # Case 2: The bucket name is already taken
+        except (Exception, BaseException) as ex:
+            logger.error(
+                "Cannot get or create {} (exit code {}):\n{}".format(
+                    bucket_name, ex.returncode, ex.output.decode()
+                )
+            )
+            log_verbose_traceback(ex)
+            raise ex
+
+        logger.debug("bucket=%s" % self.bucket.name)
+        logger.debug("subdir=%s" % self.gs_subdir)
+
+    def shutdown(self):
+        """shutdown deletes build packages if the user didn't request to clean
+           up the cache. At this point we've already cancelled running jobs.
+        """
+        # Delete build packages only if user regooglquested no cache
+        if self._save_storage_cache:
+            logger.debug("Requested to save workflow cache, skipping cleanup.")
```

Comment:
Workflow cache is misleading (could mean results instead). Rather say workflow sources?

Reply:
Done.